File: examples/java-gradle/src/main/java/com/example/impl/InMemoryProductRepository.java
package com.example.impl;

import com.example.interfaces.Repository;
import com.example.models.Product;

import java.math.BigDecimal;
import java.util.*;
import java.util.concurrent.atomic.AtomicLong;

/**
 * In-memory implementation of Product repository.
 */
public class InMemoryProductRepository implements Repository<Product> {

    private final Map<Long, Product> storage = new HashMap<>();
    private final AtomicLong idGenerator = new AtomicLong(1);

    @Override
    public Optional<Product> findById(Long id) {
        return Optional.ofNullable(storage.get(id));
    }

    @Override
    public List<Product> findAll() {
        return new ArrayList<>(storage.values());
    }

    @Override
    public Product save(Product entity) {
        Product toSave = entity.id() == null
                ? entity.withId(idGenerator.getAndIncrement())
                : entity;
        storage.put(toSave.id(), toSave);
        return toSave;
    }

    @Override
    public boolean delete(Long id) {
        return storage.remove(id) != null;
    }

    public List<Product> findInStock() {
        return storage.values().stream()
                .filter(Product::inStock)
                .toList();
    }

    public List<Product> findByPriceRange(BigDecimal min, BigDecimal max) {
        return storage.values().stream()
                .filter(p -> p.price().compareTo(min) >= 0 && p.price().compareTo(max) <= 0)
                .toList();
    }

    public List<Product> findByNameContaining(String namePart) {
        return storage.values().stream()
                .filter(p -> p.name().toLowerCase().contains(namePart.toLowerCase()))
                .toList();
    }

    public BigDecimal calculateTotalValue() {
        return storage.values().stream()
                .map(Product::price)
                .reduce(BigDecimal.ZERO, BigDecimal::add);
    }
}

File: examples/kotlin-gradle/src/main/kotlin/com/example/impl/InMemoryProductRepository.kt
package com.example.impl

import com.example.interfaces.Repository
import com.example.models.Product
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicLong

/**
 * In-memory implementation of Product repository.
 */
class InMemoryProductRepository : Repository<Product> {

    private val storage = ConcurrentHashMap<Long, Product>()
    private val idGenerator = AtomicLong(1)

    override fun findById(id: Long): Product? = storage[id]

    override fun findAll(): List<Product> = storage.values.toList()

    override fun save(entity: Product): Product {
        val toSave = if (entity.id == null) {
            entity.withId(idGenerator.getAndIncrement())
        } else {
            entity
        }
        storage[toSave.id!!] = toSave
        return toSave
    }

    override fun delete(id: Long): Boolean = storage.remove(id) != null

    fun findInStock(): List<Product> =
        storage.values.filter { it.inStock }

    fun findByPriceRange(min: Double, max: Double): List<Product> =
        storage.values.filter { it.price in min..max }

    fun findByNameContaining(namePart: String): List<Product> =
        storage.values.filter { it.name.contains(namePart, ignoreCase = true) }

    fun calculateTotalValue(): Double =
        storage.values.sumOf { it.price }
}
